### DISCUSSION TITLE
Must edge_index start at 0 when PYG constructs a graph?

### DISCUSSION THREAD

**Discussion by yaoysyao** [pyg-team/pytorch_geometric#7242] [Q&A] [answered]
When I use PYG to construct a graph for a paragraph based on the sequence of words, for example: I am a student, I set the ID for each word in the dictionary, such as I, am: 10, 15, student: a: 20, 30, when I sequence according to the structure diagram, Be based on the relationship between the edge of [10], [15, 10], [15, 30], [30], [30], [20, 30], node for [[10], [15], [20], [30]], when I use the Data appeared abnormal, when I put all of the id is amended as: When 0,1,2,3, the relation of edges also becomes [0,1],[1,0],[1,2],[2,1],[2,3],[3,2], which is found to be correct at this time. Therefore, I would like to ask, when I construct a graph, do nodes have to start from 0?
Here is my code：
x = torch.tensor([[4279], [404], [2389], [3691]], dtype=torch.float)
Correct code：
edge_index = torch.tensor([[0, 1], [0, 2], [0, 3], [1, 0], [1, 2], [2, 1], [2, 0], [2, 3], [3, 2], [3, 0]], dtype=torch.long)
error code：
edge_index = torch.tensor([[4279, 404], [4279, 2389], [4279, 3691], [404, 4279], [404, 2389],
                           [2389, 404], [2389, 4279], [2389, 3691], [3691, 2389], [3691, 4279]], dtype=torch.long)

data = Data(x=x, edge_index=edge_index.t().contiguous())
print(data)
print(data.validate(raise_on_error=True))
error message:
ValueError: 'edge_index' contains larger indices than the number of nodes (4) in 'Data' (found 4279)

**Comment by rusty1s** [comment]
Yes, `edge_index` needs to hold indices from zero to number of nodes - 1 in your graph in PyG.

**Reply by yaoysyao** [reply]
Thank you very much for your answer, because I am responsible for text classification, and I need to construct each piece of data into a graph with nodes for each word in this piece of data. In the process of data processing, I converted the ID operation for each word in all the data in the data set. The id starts from 0, and I set the id for all the words that do not repeat. My processing method is similar to the processing method in the source code of text-level-gcn, and then convert the word in each data to id. In this case, there is no way to ensure that the word id starts from 0, because the word id in each data does not always start from 0. Is there a good way to deal with this situation? I would appreciate it very much.

**Reply by rusty1s** [reply, marked as answer]
I think you are mixing up node IDs and the IDs that are used to map to words. While the `edge_index` always needs to refer to `0` to `number_of_nodes - 1` (to index the corresponding feature matrix), you can have a node ID/attribute for each node that maps to the ID of the corresponding word.

**Reply by yaoysyao** [reply]
> I think you are mixing up node IDs and the IDs that are used to map to words. While the `edge_index` always needs to refer to `0` to `number_of_nodes - 1` (to index the corresponding feature matrix), you can have a node ID/attribute for each node that maps to the ID of the corresponding word.

Yes, you are absolutely right, in fact I modified my code to do just that. As I constructed a graph for each piece of data, I reconstructed a new dictionary for the old array of word ids. The dictionary started at 0, mapped to the corresponding word ID, and then I used the new 0 starting ID as the node ID
Here are some examples of the methods I use:
Data word ID: [10,20,25,60,75,80,90]
Constructed node ids: 0,1,2,3,4,5,6
I will use the newly constructed node ID to construct a graph for each piece of data for graph classification
Thank you for your advice
